### PR TITLE
fix: custom component name should be constrained explicitely - WF-52

### LINKED ISSUE #517: custom component name should be constrained explicitely
Custom component names defined by the dev in custom_components/index.ts must contain the full original name of the component lower-case.

For example, A customized the Metric component called "customMetric", compiles but can't be dragged into the ComponentRenderer, whereas "custommetric" and "nonsensemetric" work fine. AFAIK the constraints for naming are not in the docs. video of the behavior from the invalid name is below

https://github.com/user-attachments/assets/4d9c69e6-5511-47be-90aa-d1f79447292e

### PR DESCRIPTION
**Pull request by FabienArcellier**
fix #517 

---
image

### CODE REVIEW

File: src/ui/src/custom_components/core.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	return customComponents;
+}
+
+/**
+ * Audit the validity of custom components
+ *
+ * @param key
+ * @param customComponent
+ */
+export function _auditCustomComponents(key: string, customComponent: unknown) {
```
Comment (madeindjs) [inline src/ui/src/custom_components/core.ts:22]: 1. You can remove the `_` prefix and simply not export this function
2. I think the function name could be more meaningfull
```suggestion
function checkComponentName(key: string) {
```

File: src/ui/src/custom_components/core.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+	// Checks that the key contains only alphanumeric characters and underscores without capital letters
+	// @see https://github.com/writer/writer-framework/issues/517
+	const regex = /^[a-z0-9_]+$/;
+	const is_valid_key = regex.test(key);
+	if (!is_valid_key) {
```
Comment (madeindjs) [inline src/ui/src/custom_components/core.ts:27]: In JS, we prefer camlCase like `isValidKey`.
```suggestion
	const isValidKey = regex.test(key);
	if (!isValidKey) {
```

File: src/ui/src/custom_components/core.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+		throw Error(
+			`custom component key '${key}' is not valid. custom component should be declared with only alphanumeric lowercase and _.`,
+		);
```
Comment (madeindjs) [inline src/ui/src/custom_components/core.ts:30]: Question, do we really want to stop the application here ? Do you think `console.warn` could be enough ?

File: src/ui/src/custom_components/core.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+/**
+ * Declare custom components
+ *
+ * @param customComponents list of components in key-value form
+ */
+export function declareCustomComponents(
```
Comment (madeindjs) [inline src/ui/src/custom_components/core.ts:6]: I think the function name could be more meaningfull. Maybe `checkCustomComponents`

Comment (FabienArcellier) [reply src/ui/src/custom_components/core.ts:6]: In this case I want to call the function like this `declareCustomComponents`. It seems more logical when you read the calling code :

```typescript
export default declareCustomComponents({
	bubblemessage: BubbleMessage,
	bubbleMessageadvanced: BubbleMessageAdvanced,
});
```

**Comment by ramedina86** [comment]
@madeindjs thanks for taking a look at this

@FabienArcellier I think we should check during load, it'll be more straightforward and reliable. Also, let's try to keep `core` as lean as we can.

We can intercept the keys here:

`importCustomComponentTemplate` in `loadExtensions.ts` (for loading via transpiled extension)
`if (WRITER_LIVE_CCT === "yes") {...` in `templateMap.ts` (for loading live while developing)

**Comment by FabienArcellier** [comment]
> @madeindjs thanks for taking a look at this
> 
> @FabienArcellier I think we should check during load, it'll be more straightforward and reliable. Also, let's try to keep `core` as lean as we can.
> 
> We can intercept the keys here:
> 
> `importCustomComponentTemplate` in `loadExtensions.ts` (for loading via transpiled extension) `if (WRITER_LIVE_CCT === "yes") {...` in `templateMap.ts` (for loading live while developing)

I used this approach. I am looking for a way to raise the error during the build or using a dedicated npm command.
The problem only appears in runtime when you open the web application. It does not feel great.

**Comment by FabienArcellier** [comment]
I have implemented a command `npm run custom.check` that verify this rule. It will allow to check the status without running the entire app and allow a person to implement a start of continuous integration pipeline when building custom components. I add the info in the documentation.

image

**Pull request merged**